File: project/core/module/lane/import/src/main/java/vdi/lane/imports/ImportException.java
package vdi.lane.imports;

import java.util.function.BiConsumer;

public class ImportException extends RuntimeException {
  private final long userId;
  private final String datasetId;

  public ImportException(long userId, String datasetId, Throwable cause) {
    super(String.format("import failed for dataset %d/%s", userId, datasetId), cause);
    this.userId = userId;
    this.datasetId = datasetId;
  }

  public long getUserId() {
    return userId;
  }

  public String getDatasetId() {
    return datasetId;
  }

  public void log(BiConsumer<String, Throwable> logFn) {
    logFn.accept(getMessage(), this);
  }
}

File: project/core/module/lane/import/src/main/kotlin/vdi/lane/imports/ImportException.kt
package vdi.lane.imports

import vdi.model.meta.UserID

class ImportException(
  val ownerId: UserID,
  val datasetId: String,
  message: String,
  cause: Throwable?,
): Throwable(message, cause) {
  constructor(ownerId: UserID, datasetId: String): this(
    ownerId,
    datasetId,
    "import failed for dataset $ownerId/$datasetId",
    null,
  )

  constructor(ownerId: UserID, datasetId: String, message: String): this(
    ownerId,
    datasetId,
    message,
    null
  )

  constructor(ownerId: UserID, datasetId: String, cause: Throwable): this(
    ownerId,
    datasetId,
    "import failed for dataset $ownerId/$datasetId",
    cause,
  )

  fun log(logFn: (String, Throwable) -> Unit) {
    logFn(message!!, this)
  }
}